_posts/2017-09-04-Secure-your-architecture-part1.md

```diff
@@ -0,0 +1,413 @@
+To add Redis and Spring Session to the classpath, add the following in the pom.xml
+
+{% highlight maven %}
+
+<dependency>
+    <groupId>org.springframework.session</groupId>
+	<artifactId>spring-session-data-redis</artifactId>
+	<version>(version of current release)</version>
+	<type>pom</type>
+</dependency>
+<dependency>
+	<groupId>org.springframework.boot</groupId>
+	<artifactId>spring-boot-starter-data-redis</artifactId>
+</dependency>
+
+{% endhighlight %}
+
+### Enabling Spring Session & Redis
+Just one annotation is needed to enable it with default configuration on your application.
+If you need to customize your Redis configuration, you can opt in one of the following configurations:
+* [Sentinel configuration](http://docs.spring.io/spring-data/redis/docs/current/reference/html/#redis:sentinel)
+* [Default configuration](https://docs.spring.io/spring-session/docs/current/reference/html5/guides/boot.html#boot-redis-configuration)
+
+{% highlight java %}
+
+@EnableRedisHttpSession
+public class Application {
+}
+{% endhighlight %}
+
+### Single Sign On Flow
+
+<div class="row" style="margin: 0 auto 2.5rem auto; width: 100%;">
+  <div class="col-md-offset-3 col-md-6" style="padding: 0;">
+	{% include image.html img="/img/microservices/part1/SSOSecurity.jpg" alt="Single Sign On" title="SSO" %}
+</div>
+</div>
+
+* The user logs into the UAA server and grants permission with frontend 1
+* The UAA server creates a sessionID for this user
+* The redis database confirms the creation
+* After the redirect mechanism of the authorization grant, the UAA creates a cookie in the user’s browser with the user’s sessionID and returns a JWT token specific for that frontend
+* The frontend 2 uses the same browser (same sessionID cookie) to visit the UAA server
+* The UAA finds a sessionID in the request and consults the redis database for its existence
+* The redis database confirms the existence
+* The UAA directly logs in frontend 2 and returns a JWT token specific to frontend 2
+
+
+
+# Secure your microservice
+When enabling security in your service, the most common issues are developer-induced.
+Either there is a lack of built-in or easy security controls or we make trade-offs for functionality/features over security.
+Still, we have to think about who can access this functionality and what they can do with it. 
+We enter the phase where we passed the authentication and retrieved a JSON Web Token from our zuul.
+We are in a 'downstream service', where data is load balanced from zuul. 
+Next thing to question is, how do we decode this JWT? How can we secure our classes and methods with the help of Spring Security?
+
+## Decoding the JWT
+It is the responsibility of a microservice (Resource Server) to extract information about the user and client application from the JWT token and make an access decision based on that information.
+With the help of a JWT decoder, it extracts the users information and put it in the security context.
```

Comment:
Decoding the JWT allows extraction of the users information and allows putting it in